### PR TITLE
Improve mechanism for getting final task stats

### PR DESCRIPTION
**Pull request by nileema**

### CODE REVIEW

**Comment by nileema** [comment]
@dain - some of the commits can be squashed together, I left them separate because I thought it would help with the review.

File: presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java [hunk 1]
```diff
@@ -611,6 +602,15 @@ void failTask(Throwable cause)
         abort(failWith(getTaskStatus(), FAILED, ImmutableList.of(toFailure(cause))));
     }
 
+    private HttpUriBuilder getHttpUriBuilder(URI uri)
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java]: I'd change this to `uriBuilder(TaskStatus)`

File: presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java [hunk 2]
```diff
@@ -403,6 +402,12 @@ private synchronized void updateTaskInfo(TaskInfo newValue, List<TaskSource> sou
         partitionedSplitCountTracker.setPartitionedSplitCount(getPartitionedSplitCount());
     }
 
+    private void updateTaskInfo(TaskInfo taskInfo)
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java]: I like the name change from `updateTaskInfo` to `processTaskUpdate`, but I don't understand why this method was extracted.  It doesn't seem to be used anywhere except `processTaskUpdate`.

File: presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java [hunk 3]
```diff
@@ -214,8 +214,10 @@ public void success(TaskInfo newValue)
     {
         try (SetThreadName ignored = new SetThreadName("TaskInfoFetcher-%s", taskId)) {
             lastUpdateNanos.set(System.nanoTime());
-            updateStats(currentRequestStartNanos.get());
 
+            synchronized (this) {
+                updateStats(currentRequestStartNanos.get());
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java]: I believe the synchronization should be inside of `updateStats`.  Specifically:
```java
        private void updateStats(long currentRequestStartNanos)
        {
            long roundTripMillis = Duration.nanosSince(currentRequestStartNanos).toMillis();
            synchronized (this) {
                stats.updateRoundTripMillis(roundTripMillis);
            }
        }

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java]: Oh I see, I had this wrong.  You are trying to protect `currentRequestStartNanos.get()`, so ignore my comment above. 

In this case, I think we want to minimize the scope of the `synchronized` to the  `get` call and not include all of the expensive calls for `Duration` and updating the stats.  So maybe code like this:
```java
long startNanos;
synchronized (this) {
    startNanos = currentRequestStartNanos.get();
}
updateStats(startNanos);
```

Or even better add a `private synchronized` method to get the `currentRequestStartNanos`

Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java]: BTW, this is already fixed in trunk

File: presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java (hunk 2 above)
Comment (nileema) [inline presto-main/src/main/java/com/facebook/presto/server/remotetask/HttpRemoteTask.java]: It is used in the next commit.

**Comment by dain** [comment]
Assuming travis passes, looks good

**Pull request closed without merging**